### ISSUE TITLE
High RAM when the garbage collection of frame-objects is deferred

### ISSUE THREAD

**Issue by cid-chan** [vapoursynth/vapoursynth#794]
Well, now I get huge memory spikes.  IIRC, high RAM usage was the reason for these `gc.collect()` calls.

<sub><sub>Also, I like how off my initial diagnosis of the matter was.  Heh, can't win them all.</sub></sub>

_Originally posted by @Kamekameha in https://github.com/vapoursynth/vapoursynth/issues/792#issuecomment-930597823_

-----

All memory-issues relating to Frame-Instances are almost textbook-examples of python resource-management problems.

RAM are a resource, but usually not that problematic in python applications. For larger applications, file-descriptors are the larger issue, so I will explain the problem using file-descriptors, but the issue the same as the RAM usage of frame-instances.

This stack-overflow answer (https://stackoverflow.com/a/49513137) describes a general problem with garbage collectors -- they do not promise to immediately call a destructor when nobody has a reference on them anymore. Instead they promise to collect them at some point. The reference counting is in fact an implementation detail of CPython many applications unknowingly rely on (see the generator example of https://www.pypy.org/compat.html). This is made worse by cython generating pypy-compatible binaries.

The canonical answer is to allow resources to be closed manually. Resources that have not been closed by the time they are collection will, of course, continue to be closed.

**Here is my proposal:**
We add a `[Video|Audio]Frame.close()`-method and add a context-manager for these frames so users can just do this and never forget to close the resource then:
```py
with clip.get_frame(0) as f:
    ...
```

We add some additional parameter (e.g. `autoclose=False`) to `.frame()` to automatically close the frame after each iteration. The default value is chosen to retain the current behavior of vapoursynth.

`node.output(x)` will then use the above parameter to fix the memory-issues.

**A note on ModifyFrame**
Usually all frames inside the function will not escape the current stack frame. Nor will the stack frame ever be persisted. It should (on CPython at least) continue to be collected then the function call returns to vapoursynth.

Because of this use of frames, we cannot bind a runtime warning to frames that are not manually released at the time of collection. On the other hand, I do not think printing a runtime-warning is neccessary.

**Comment by myrsloik** [comment]
Makes sense. Let's do it.

**Comment by Kamekameha** [comment]
Nah, here's your reference cycle: https://github.com/vapoursynth/vapoursynth/blob/a22068fd0e36a9a2a5c4a6769830868f841b58e5/src/cython/vapoursynth.pyx#L646

In https://github.com/vapoursynth/vapoursynth/issues/785#issuecomment-927202707 when I "confirmed" the source of the problem I wrote the shim of code I used from scratch, so I failed to notice both the ref-cycle and the `gc.collect()` call to mitigate it.

I finally took `CallbackData` seriously and looked into it to locate the aforementioned cycle.

**Comment by Kamekameha** [comment]
So with something along the lines of this one can avoid creating said cycle:
```diff
diff --git a/src/cython/vapoursynth.pyx b/src/cython/vapoursynth.pyx
index db6caed2..8b7cb7fe 100644
--- a/src/cython/vapoursynth.pyx
+++ b/src/cython/vapoursynth.pyx
@@ -43,6 +43,7 @@ import weakref
 import atexit
 import contextlib
 import logging
+import functools
 from threading import local as ThreadLocal, Lock, RLock
 from types import MappingProxyType
 from collections import namedtuple
@@ -643,18 +644,20 @@ cdef class CallbackData(object):
     def for_future(self, object future, object wrap_call=None):
         if wrap_call is None:
             wrap_call = lambda func, *args, **kwargs: func(*args, **kwargs)
-        self.callback = self.handle_future
+        self.callback = functools.partial(CallbackData.handle_future,
+                                          self.env, future, wrap_call)
         self.future = future
         self.wrap_cb = wrap_call
 
-    def handle_future(self, node, n, result):
+    @staticmethod
+    def handle_future(env, future, wrapper, node, n, result):
         if isinstance(result, Error):
-            func = self.future.set_exception
+            func = future.set_exception
         else:
-            func = self.future.set_result
+            func = future.set_result
 
-        with use_environment(self.env).use():
-            self.wrap_cb(func, result)
+        with use_environment(env).use():
+            wrapper(func, result)
 
     def receive(self, n, result):
         self.callback(self.node, n, result)
```

**Comment by DJATOM** [comment]
With that patch I'm getting `0.6760003566741943` with BlankClip (script from #785).

**Comment by Kamekameha** [comment]
Is that good news?  Does it also keep memory at bay for you?

**Comment by cid-chan** [comment]
> So with something along the lines of this one can avoid creating said cycle:
> 
> ```diff
> diff --git a/src/cython/vapoursynth.pyx b/src/cython/vapoursynth.pyx
> index db6caed2..8b7cb7fe 100644
> --- a/src/cython/vapoursynth.pyx
> +++ b/src/cython/vapoursynth.pyx
> @@ -43,6 +43,7 @@ import weakref
>  import atexit
>  import contextlib
>  import logging
> +import functools
>  from threading import local as ThreadLocal, Lock, RLock
>  from types import MappingProxyType
>  from collections import namedtuple
> @@ -643,18 +644,20 @@ cdef class CallbackData(object):
>      def for_future(self, object future, object wrap_call=None):
>          if wrap_call is None:
>              wrap_call = lambda func, *args, **kwargs: func(*args, **kwargs)
> -        self.callback = self.handle_future
> +        self.callback = functools.partial(CallbackData.handle_future,
> +                                          self.env, future, wrap_call)
>          self.future = future
>          self.wrap_cb = wrap_call
>  
> -    def handle_future(self, node, n, result):
> +    @staticmethod
> +    def handle_future(env, future, wrapper, node, n, result):
>          if isinstance(result, Error):
> -            func = self.future.set_exception
> +            func = future.set_exception
>          else:
> -            func = self.future.set_result
> +            func = future.set_result
>  
> -        with use_environment(self.env).use():
> -            self.wrap_cb(func, result)
> +        with use_environment(env).use():
> +            wrapper(func, result)
>  
>      def receive(self, n, result):
>          self.callback(self.node, n, result)
> ```

Awesome. Can you make a PR with that change so we can discuss it separately? I think your example proves my initial point: A reference cycle is far too easy to generate accidentally in larger applications.

**Comment by DJATOM** [comment]
> Is that good news? Does it also keep memory at bay for you?

Yeah, I think it's good (was 1.14 sec before). Speedup is noticeable.

**Comment by Kamekameha** [comment]
Okay, nice.

> Can you make a PR with that change so we can discuss it separately?

Sure, give me a moment.

> P.S. I feel blind rn, Kamekameha.

I guess that makes two of us.

**Comment by myrsloik** [comment]
So this is fixed without manually releasing frames I guess?
